### ISSUE TITLE
Exception support

### ISSUE THREAD

**Issue by carlokok** [bytecodealliance/wasmtime#2049]
#### Feature

Cranelift currently doesn't support SEH (but has some support for finally I think). This is a feature request to support properly exception (landingpads on Posix, SEH on windows, maybe SLJL for others)

#### Benefit

WASM has plans to support exceptions, at which point it's going to be needed. Win64 has partial SEH support (iirc it just emits what's needed, nothing more), this would allow for c++ /java/.net exception support.

#### Implementation

Tricky one. I know LLVM has two alternative ways of doing this. I think with that in mind it should be possible to do a single implementation. What's basically needed is a way to say:
from here
[code]

till here is a protected block, on exceptions go to X, for finalization go to Y. For the handler itself it should be possible to define the personality to call.

Throw, Rethrow etc can all probably be done on a library level.

#### Alternatives

The only alternative is to not do it.

**Comment by bjorn3** [comment]
Duplicate of https://github.com/bytecodealliance/wasmtime/issues/1677

**Comment by carlokok** [comment]
I agree it's related, but unwind and catch are slightly distinct actions and your issue doesn't cover that?

**Comment by bjorn3** [comment]
Catch is implemented by returning `_URC_HANDLER_FOUND` instead of `_URC_INSTALL_CONTEXT` from the personality function and then jumping from the landing pad to the block handling the catch instead of calling `_Unwind_Resume` like for cleanup.

**Comment by bjorn3** [comment]
Given the above comments I think this can be closed as duplicate.

**Comment by carlokok** [comment]
I missed your reply last year but does that cover windows at all? Windows doesn't use personalities? At least not like that?

**Comment by bjorn3** [comment]
SEH seems to use exception handlers and termination handlers. I think catching an exception instead of doing cleanup is done by using an exception handler instead of termination handler. This is based on a cursory reading of the docs though, so I may be wrong. In any case both catching exceptions and cleanup needs the same codegen handling to support an alternative location to which a call can "return".

**Comment by cfallin** [comment]
Cranelift supports exceptions now, and they are used by Wasmtime exception handling and `cg_clif` panic-unwind support; closing issue.
